resources/aws_ec2.go

```diff
@@ -0,0 +1,602 @@
+						}:
+						case <-closeChan:
+							return
+						}
+					}
+				}
+			}
+			if obj.State == "terminated" {
+				obj.client.WaitUntilInstanceExistsWithContext(ctx, diInput)
+				stateOutput, err := obj.client.DescribeInstances(diInput)
+				if err != nil {
+					select {
+					case awsChan <- &chanStruct{
+						err: errwrap.Wrapf(err, "error describing instances"),
+					}:
+					case <-closeChan:
+					}
+					return
+				}
+				if len(stateOutput.Reservations) == 1 {
+					{
+						select {
+						case awsChan <- &chanStruct{
+							str: "exists",
+						}:
+						case <-closeChan:
+							return
+						}
+					}
+				}
+			}
+			select {
+			case <-closeChan:
+				return
+			default:
+			}
+		}
+	}()
+	for {
+		select {
+		case event := <-obj.Events():
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				close(closeChan)
+				return *exit
+			}
+		case msg, ok := <-awsChan:
+			if !ok {
+				return *exit
+			}
+			if err := msg.err; err != nil {
+				return err
+			}
+			log.Printf("%s: State: %s", obj, msg.str)
+			obj.StateOK(false)
+			send = true
+		}
+		if send {
+			send = false
+			obj.Event()
+		}
+	}
+}
+
+// CheckApply method for AwsEc2 resource.
+func (obj *AwsEc2Res) CheckApply(apply bool) (checkOK bool, err error) {
+	log.Printf("%s: CheckApply(%t)", obj, apply)
+
+	const waitTimeout = 400
+
+	diInput := ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []*string{aws.String(obj.prependName())},
+			},
+			{
+				Name: aws.String("instance-state-name"),
+				Values: []*string{
+					aws.String("running"),
+					aws.String("pending"),
+					aws.String("stopped"),
+					aws.String("stopping"),
+				},
+			},
+		},
+	}
+	diOutput, err := obj.client.DescribeInstances(&diInput)
+	if err != nil {
+		return false, errwrap.Wrapf(err, "error describing instances")
+	}
+
+	if len(diOutput.Reservations) < 1 && obj.State == "terminated" {
+		return true, nil
+	}
+	if len(diOutput.Reservations) == 1 && *diOutput.Reservations[0].Instances[0].State.Name == obj.State {
+		return true, nil
+	}
+	if !apply {
+		return false, nil
+	}
+
+	if len(diOutput.Reservations) > 1 {
+		return false, fmt.Errorf("too many reservations")
+	}
+	ctx, cancel := context.WithTimeout(context.TODO(), waitTimeout*time.Second)
+	defer cancel()
+	if len(diOutput.Reservations) == 1 {
+		instanceID := diOutput.Reservations[0].Instances[0].InstanceId
+		describeInput := &ec2.DescribeInstancesInput{
+			InstanceIds: []*string{instanceID},
+		}
+		if len(diOutput.Reservations[0].Instances) > 1 {
+			return false, fmt.Errorf("more than one instance was returned")
+		}
+		if obj.State == "running" {
+			startInput := &ec2.StartInstancesInput{
+				InstanceIds: []*string{instanceID},
+			}
+			_, err := obj.client.StartInstances(startInput)
+			if err != nil {
+				return false, errwrap.Wrapf(err, "error starting instance")
+			}
+			obj.client.WaitUntilInstanceRunningWithContext(ctx, describeInput)
```

Comment:
This is a clever, but unfortunately not correct in certain cases. Better to check the error, and you can also test if it's of type context whatever eg: https://golang.org/pkg/context/#pkg-variables or nil or a regular error unrelated to the context.